### PR TITLE
Add LocationCompleter to Compute cmdlets

### PR DESCRIPTION
**Pull request by maddieclayton**
## Description
Add the Location completer to the -Location parameter for all Compute cmdlets with the correct providers.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

File: src/ResourceManager/Compute/Commands.Compute/VirtualMachineSizes/GetAzureVMSizeCommand.cs [hunk 1]
```diff
@@ -35,7 +36,7 @@ public class GetAzureVMSizeCommand : VirtualMachineSizeBaseCmdlet
            Position = 0,
            ParameterSetName = ListVirtualMachineSizeParamSet,
            ValueFromPipelineByPropertyName = true,
-           HelpMessage = "The location name.")]
+           HelpMessage = "The location name."), LocationCompleter(new string[] { "Microsoft.Compute/locations/vmSizes" })]
```
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/VirtualMachineSizes/GetAzureVMSizeCommand.cs]: Move LocationCompleter() to [ValidateNotNullOrEmpty] like NewAzureVMCommand.cs

File: src/ResourceManager/Compute/Commands.Compute/Usage/GetAzureVMUsageCommand.cs [hunk 2]
```diff
@@ -30,7 +31,7 @@ public class GetAzureVMUsageCommand : VirtualMachineUsageBaseCmdlet
            Mandatory = true,
            Position = 0,
            ValueFromPipelineByPropertyName = true,
-           HelpMessage = "The location name.")]
+           HelpMessage = "The location name."), LocationCompleter(new string[] { "Microsoft.Compute/locations/usages" })]
```
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/Usage/GetAzureVMUsageCommand.cs]: Move LocationCompleter() to [ValidateNotNullOrEmpty] like NewAzureVMCommand.cs

File: src/ResourceManager/Compute/Commands.Compute/AvailabilitySets/NewAzureAvailabilitySetCommand.cs [hunk 3]
```diff
@@ -46,7 +47,7 @@ public class NewAzureAvailabilitySetCommand : AvailabilitySetBaseCmdlet
             Mandatory = true,
             Position = 2,
             ValueFromPipelineByPropertyName = true,
-            HelpMessage = "The location.")]
+            HelpMessage = "The location."), LocationCompleter(new string[] { "Microsoft.Compute/availabilitySets" })]
```
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/AvailabilitySets/NewAzureAvailabilitySetCommand.cs]: Move LocationCompleter() to [ValidateNotNullOrEmpty] like NewAzureVMCommand.cs

File: src/ResourceManager/Compute/Commands.Compute/Extension/DSC/SetAzureVMDscExtensionCommand.cs [hunk 4]
```diff
@@ -170,7 +171,7 @@ public class SetAzureVMDscExtensionCommand : VirtualMachineExtensionBaseCmdlet
         [Parameter(
             Mandatory = false,
             ValueFromPipelineByPropertyName = true,
-            HelpMessage = "Location of the resource.")]
+            HelpMessage = "Location of the resource."), LocationCompleter(new string[] { "Microsoft.Storage/storageAccounts" })]
```
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/Extension/DSC/SetAzureVMDscExtensionCommand.cs]: Move LocationCompleter() to [ValidateNotNullOrEmpty] like NewAzureVMCommand.cs

File: src/ResourceManager/Compute/Commands.Compute/Extension/Diagnostics/SetAzureRmVMDiagnosticsExtension.cs [hunk 5]
```diff
@@ -104,7 +105,7 @@ public string DiagnosticsConfigurationPath
             Mandatory = false,
             Position = 7,
             ValueFromPipelineByPropertyName = true,
-            HelpMessage = "The location.")]
+            HelpMessage = "The location."), LocationCompleter(new string[] { "Microsoft.Storage/storageAccounts" })]
```
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/Extension/Diagnostics/SetAzureRmVMDiagnosticsExtension.cs]: Move LocationCompleter() to [ValidateNotNullOrEmpty] like NewAzureVMCommand.cs

File: src/ResourceManager/Compute/Commands.Compute/Extension/SetAzureVMExtensionBaseCmdlet.cs [hunk 6]
```diff
@@ -54,7 +55,7 @@ public class SetAzureVMExtensionBaseCmdlet : VirtualMachineExtensionBaseCmdlet
         [Parameter(
             Mandatory = false,
             ValueFromPipelineByPropertyName = true,
-            HelpMessage = "The location.")]
+            HelpMessage = "The location."), LocationCompleter(new string[] { "Microsoft.Compute/virtualMachines" })]
```
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/Extension/SetAzureVMExtensionBaseCmdlet.cs]: Move LocationCompleter() to [ValidateNotNullOrEmpty] like NewAzureVMCommand.cs

File: src/ResourceManager/Compute/Commands.Compute/Extension/SqlServer/SetAzureVMSqlServerExtensionCommand.cs [hunk 7]
```diff
@@ -93,7 +94,7 @@ public class SetAzureSqlServerExtensionCommand : VirtualMachineExtensionBaseCmdl
             Mandatory = false,
             Position = 8,
             ValueFromPipelineByPropertyName = true,
-            HelpMessage = "Location of the resource.")]
+            HelpMessage = "Location of the resource."), LocationCompleter(new string[] { "Microsoft.Compute/virtualMachines" })]
```
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/Extension/SqlServer/SetAzureVMSqlServerExtensionCommand.cs]: Move LocationCompleter() to [ValidateNotNullOrEmpty] like NewAzureVMCommand.cs

File: src/ResourceManager/Compute/Commands.Compute/Images/GetAzureVMImageCommand.cs [hunk 8]
```diff
@@ -37,7 +38,7 @@ public class GetAzureVMImageCommand : VirtualMachineImageBaseCmdlet
             ValueFromPipelineByPropertyName = true)]
         [Parameter(ParameterSetName = GetVMImageDetailParamSetName,
             Mandatory = true,
-            ValueFromPipelineByPropertyName = true)]
+            ValueFromPipelineByPropertyName = true), LocationCompleter(new string[] { "Microsoft.Compute/locations/publishers" })]
```
Comment (hyonholee) [inline src/ResourceManager/Compute/Commands.Compute/Images/GetAzureVMImageCommand.cs]: Move LocationCompleter() to [ValidateNotNullOrEmpty] like NewAzureVMCommand.cs

File: src/ResourceManager/Compute/Commands.Compute/Extension/Diagnostics/SetAzureRmVMDiagnosticsExtension.cs (hunk 5 above)
Comment (maddieclayton) [reply src/ResourceManager/Compute/Commands.Compute/Extension/Diagnostics/SetAzureRmVMDiagnosticsExtension.cs]: No [ValidateNotNullOrEmpty] attribute here, so I moved the LocationCompleter out on its own.

File: src/ResourceManager/Compute/Commands.Compute/AvailabilitySets/NewAzureAvailabilitySetCommand.cs (hunk 3 above)
Comment (maddieclayton) [reply src/ResourceManager/Compute/Commands.Compute/AvailabilitySets/NewAzureAvailabilitySetCommand.cs]: I moved all of the LocationCompleter attributes out to a new set of brackets for uniformity - let me know if this looks good to you.

**Comment by hyonholee** [comment]
Thanks, LGTM.

**Comment by maddieclayton** [comment]
Optimistic on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/32/

**Pull request merged**